src/replication.c

```diff
@@ -2160,136 +2182,131 @@ void syncWithMaster(connection *conn) {
+                "capa","eof","capa","psync2",NULL);
+        if (err) goto write_error;
+
+        server.repl_state = REPL_STATE_RECEIVE_AUTH;
+        return;
     }
 
+    if (server.repl_state == REPL_STATE_RECEIVE_AUTH && !server.masterauth)
+        server.repl_state = REPL_STATE_RECEIVE_PORT;
+
     /* Receive AUTH reply. */
     if (server.repl_state == REPL_STATE_RECEIVE_AUTH) {
-        err = sendSynchronousCommand(SYNC_CMD_READ,conn,NULL);
+        err = receiveSynchronousResponse(conn);
         if (err[0] == '-') {
             serverLog(LL_WARNING,"Unable to AUTH to MASTER: %s",err);
-            sdsfree(err);
+            sdsfree(err), err = NULL;
             goto error;
         }
-        sdsfree(err);
-        server.repl_state = REPL_STATE_SEND_PORT;
-    }
-
-    /* Set the slave port, so that Master's INFO command can list the
-     * slave listening port correctly. */
-    if (server.repl_state == REPL_STATE_SEND_PORT) {
-        int port;
-        if (server.slave_announce_port) port = server.slave_announce_port;
-        else if (server.tls_replication && server.tls_port) port = server.tls_port;
-        else port = server.port;
-        sds portstr = sdsfromlonglong(port);
-        err = sendSynchronousCommand(SYNC_CMD_WRITE,conn,"REPLCONF",
-                "listening-port",portstr, NULL);
-        sdsfree(portstr);
-        if (err) goto write_error;
-        sdsfree(err);
+        sdsfree(err), err = NULL;
         server.repl_state = REPL_STATE_RECEIVE_PORT;
         return;
     }
 
     /* Receive REPLCONF listening-port reply. */
     if (server.repl_state == REPL_STATE_RECEIVE_PORT) {
-        err = sendSynchronousCommand(SYNC_CMD_READ,conn,NULL);
+        err = receiveSynchronousResponse(conn);
         /* Ignore the error if any, not all the Redis versions support
          * REPLCONF listening-port. */
         if (err[0] == '-') {
             serverLog(LL_NOTICE,"(Non critical) Master does not understand "
                                 "REPLCONF listening-port: %s", err);
         }
-        sdsfree(err);
-        server.repl_state = REPL_STATE_SEND_IP;
-    }
-
-    /* Skip REPLCONF ip-address if there is no slave-announce-ip option set. */
-    if (server.repl_state == REPL_STATE_SEND_IP &&
-        server.slave_announce_ip == NULL)
-    {
-            server.repl_state = REPL_STATE_SEND_CAPA;
-    }
-
-    /* Set the slave ip, so that Master's INFO command can list the
-     * slave IP address port correctly in case of port forwarding or NAT. */
-    if (server.repl_state == REPL_STATE_SEND_IP) {
-        err = sendSynchronousCommand(SYNC_CMD_WRITE,conn,"REPLCONF",
-                "ip-address",server.slave_announce_ip, NULL);
-        if (err) goto write_error;
-        sdsfree(err);
+        sdsfree(err), err = NULL;
         server.repl_state = REPL_STATE_RECEIVE_IP;
         return;
     }
 
+    if (server.repl_state == REPL_STATE_RECEIVE_IP && !server.slave_announce_ip)
```

Comment:
This introduces a regression, in case slave_announce_ip changes in-between. Not sure if it's worth the trouble storing extra state but just saying.

Reply:
i agree, i posted about this concern in my first post of the PR, asking for advise of how this can be resolved in a cleaner manner.
one possibility is to store flags on what we sent in a function static variable.
do you have a better idea, or a preference?

Reply:
Oh sorry I skipped that comment. A static will be ugly but I think it's good enough, I think ideally when (if?) we do refactor this we can maintain a dynamic list of callbacks to handle expected replies which we update when sending a command.